File: InvokerFunction/go/first-arch-go/functions/log/handler_log.go
package main

import (
	"encoding/json"
	"fmt"
	"function/injector"
	"log"
	"os"
	"time"

	"github.com/nats-io/nats.go"
	"github.com/sirupsen/logrus"
)

var logger = logrus.New()

func setup() {

	logger.SetFormatter(&logrus.JSONFormatter{TimestampFormat: "2006-01-02T15:04:05.000Z07:00"})
	file, err := os.OpenFile("./logs.txt", os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0666)
	if err == nil {
		logger.SetOutput(file)
	} else {
		log.Fatal(err)
	}
}

func handler() {
	i := injector.NewInjector()
	aclService := i.GetServiceById("log")

	topic := aclService.ServiceTopic
	message := map[string]string{"timestamp": "2024-11-28T16:05:34", "message": "ciao", "severity": "info"}
	result, err := invokeFunction(topic, message)
	if err != nil {
		log.Fatal(err)
	}
	os.Stdout.WriteString("logging function result: " + result)
}

func invokeFunction(topic string, message map[string]string) (string, error) {
	nc, err := nats.Connect(nats.DefaultURL)
	if err != nil {
		return "error connecting to NATS", fmt.Errorf("error connecting to NATS: %w", err)
	}
	defer nc.Close()

	start := time.Now()
	msg, err := json.Marshal(message)
	if err != nil {
		return "error marshalling message", fmt.Errorf("error marshaling message: %w", err)
	}

	response, err := nc.Request(topic, msg, nats.DefaultTimeout)
	if err != nil {
		return "error publishing message", fmt.Errorf("error publishing message: %w", err)
	}
	end := time.Now()
	logger.Infof("logging function executed in %v ms", end.Sub(start).Milliseconds())

	/*
		var result string
		if err := json.Unmarshal(response.Data, &result); err != nil {
			return "failed to decode result", fmt.Errorf("error unmarshaling response: %w", err)
		}
	*/
	return string(response.Data), nil
}

func main() {
	setup()
	handler()
}

File: first-architecture/InvokerJSFunction/javascript/first-arch-js/handler_log.js
const Injector = require("./Injector.js")
const { connect } = require('nats');
const winston = require("winston");


natsURL = process.env.NATSSERVER || "nats://192.168.17.118:4222"


const logger = winston.createLogger({
    level: "info",
    format: winston.format.combine(
      winston.format.timestamp(),
      winston.format.json()
    ),
    transports: [
      new winston.transports.File({ filename: "logs.txt" }),
    ],
  });

async function hanlder(){
const injector = new Injector()
const log_service = await injector.getServiceById("log")
const topic = log_service.ServiceTopic
const message = {timestamp: "2024-11-28T16:05:34", message: "ciao", severity: "info"}
const result = await invokeFunction(topic, message)
console.log("logging result:", result)
process.exit(0, "logging function executed")
}



async function invokeFunction (topic, message){
    try {
        // Connect to the NATS server
        const nc = await connect({ servers: natsURL }); 
        const start = Date.now()
        // Publish the message
        const response = await nc.request(topic, JSON.stringify(message));
        const end = Date.now()
        logger.info(`logging function executed in ${end-start} ms `)
        // Close the connection
        nc.drain();
        return (response.data.toString())
      } catch (err) {
        console.error("Error publishing message:", err);
      }
}

hanlder()